lib/spack/spack/util/compression.py

```diff
@@ -278,6 +275,208 @@ def decompressor_for(path, ext):
+    @classmethod
+    def magic_number(cls):
+        return [cls._GNU_MAGIC, cls._POSIX_MAGIC]
+
+    @classmethod
+    def header_size(cls):
+        return max(len(cls._GNU_MAGIC), len(cls._POSIX_MAGIC))
+
+    @classmethod
+    def _bytes_check(cls, magic_bytes):
+        for magic in cls.magic_number():
+            if magic_bytes.startswith(magic):
+                return True
+        return False
+
+
+class ZipFleType(FileType):
+    _MAGIC_NUMBER = b'PK\003\004'
+    _ext = 'zip'
+
+
+VALID_FILETYPES = [BZipFileType, ZCompressedFileType,
+                   GZipFileType, LzmaFileType, TarFileType,
+                   ZipFleType]
+
+
+def extension_from_stream(stream, decompress=False):
+    """Return extension represented by stream corresponding to archive file
+    If stream does not represent an archive type recongized by Spack
+    (see `spack.util.compression.ALLOWED_ARCHIVE_TYPES`) method will return None
+
+    Extension type is derived by searching for identifying bytes
+    in file stream.
+
+    Args:
+        stream (IOBase): stream representing a file on system
+        decompress (bool) : if True, compressed files are checked
+                            for archive types beneath compression i.e. tar.gz
+                            default is False, otherwise, return top level type i.e. gz
+
+    Return:
+        extension (str|None): A string represting corresponding archive extension
+            or None as relevant.
+
+    """
+    for arc_type in VALID_FILETYPES:
+        if arc_type.is_file_of_type(stream):
+            suffix_ext = arc_type.extension()
+            prefix_ext = ''
+            if arc_type.is_compression_type() and decompress:
+                # stream represents compressed file
+                # get decompressed stream (if possible)
+                decomp_stream = arc_type.decomp_in_memory(stream)
+                if not decomp_stream:
+                    # we were unable to decompress, fallback on
+                    # filename parsing to derive extension
+                    return extension(stream.name)
+                prefix_ext = extension_from_stream(decomp_stream, decompress=decompress)
+            return suffix_ext if not prefix_ext else '.'.join([prefix_ext, suffix_ext])
+    return None
+
+
+def extension_from_file(file, decompress=False):
+    """Takes filepath representing file on system.
+    Opens file into memory and checks for spack supported archive types.
+
+    Args:
+        file (os.PathLike)
+    """
+    if os.path.exists(file):
+        with open(file, 'rb') as f:
+            ext = extension_from_stream(f, decompress)
+            # based on magic number, file is compressed
+            # tar archive. Check to see if file is abbreviated as
+            # t[xz|gz|bz2|bz]
+            if ext and ext.startswith('tar.'):
+                suf = ext.split('.')[1]
+                abbr = 't' + suf
+                if check_extension(file, abbr):
+                    return abbr
```

Comment:
This looks like it's trying to be careful to say `.tgz` if the file name has `.tgz` in it, but I'm not sure I understand why we're trying to be careful about that: I would think we could always return `.tar.gz` for files even if they are named `.tgz` (in other words, magic number parsing can dictate the extension type and doesn't have to worry about what the filename is).

Reply:
I don't think this extra check detracts from the quality of the method (if I'm wrong here, please correct me and disregard the rest of this comment) and it allows behavior to be consistent between our `extension` and `extension_from_file` methods as well as before and after this change within Spack. For the purposes of `decompressor_for`, yes, this is a frivolous distinction as the decomp strategy will be the same for `.tgz` and `tar.gz`, but in other contexts, having the appropriate string returned could be very useful (url parsing for example).

Reply:
The primary issue is that I found this confusing: I think that could be addressed by updating the docstring to explain it will try to return `.tgz` if the path contains `.tgz` (and that this might be useful for functions which want to take advantage of magic number matching to aid their parsing).

I also mentioned in `check_extension` that right now that is the same as `abbr in file`: while I think that is generally safe if `abbr` were `".tgz"`, I'm concerned that `"tgz"` (i.e. without the `"."`) is too loose of a match.

Reply:
I believe this has been addressed by the method doc string. If this is not the case, I can improve the docstring or we can discuss alternatives.